examples/haiku_simple.py

```diff
@@ -0,0 +1,164 @@
+"""
+Optuna example that optimizes a neural network classifier configuration for the
+MNIST dataset using Jax and Haiku.
+
+In this example, we optimize the validation accuracy of MNIST classification using
+jax nad haiku. We optimize the number of linear layers and learning rate of the optimizer.
+
+The example code is based on https://github.com/deepmind/dm-haiku/blob/master/examples/mnist.py
+"""
+
+from typing import Any
+from typing import Generator
+from typing import Mapping
+from typing import Tuple
+
+import haiku as hk
+import jax
+import jax.numpy as jnp
+import numpy as np
+import optax
+import tensorflow_datasets as tfds
+
+import optuna
+
+
+OptState = Any
+Batch = Mapping[str, np.ndarray]
+
+BATCH_SIZE = 128
+TRAIN_STEPS = 1000
+N_TRAIN_SAMPLES = 3000
+N_VALID_SAMPLES = 1000
+
+
+def load_dataset(
+    split: str,
+    *,
+    is_training: bool,
+    batch_size: int,
+    sample_size: int,
+) -> Generator[Batch, None, None]:
+    """Loads the sub-sampled dataset as a generator of batches."""
+    ds = tfds.load("mnist:3.*.*", split=split).take(sample_size).cache().repeat()
```

Comment:
I suspect this is inappropriate because `take` just returns the first `count` samples and so the training dataset is the first 3000 samples, the validation dataset is composed of the first 3000 samples, and the test dataset is the first 1000 samples. This would be a terrible leakage. With `skip` function, I think we can do what I think desirable, i.e., the first 3000 samples for training, following 3000 samples for validation, and the following 1000 samples for testing.

Reply:
I would like to clarify the roles of datasets
- `train` is the training dataset.
- `train_eval` is the same as training dataset but large batch size to compute evaluation metric on training dataset faster. This comes from [the original example code](https://github.com/deepmind/dm-haiku/blob/master/examples/mnist.py). 
- `test_eval` is the validation dataset like another example such as [`keras_simple.py`](https://github.com/optuna/optuna/blob/master/examples/keras_simple.py). 

Probably, I should have changed the variable name of `test_eval`. In addition, should I remove `train_eval`?

Reply:
Ah, sorry, I missed the "is_training" argument. thanks. `train_eval` being the same as `train` isn't that problematic and kind of common. Thank you for the clarification. :+1: